docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeFeature
+{
+    public const string Async = nameof(Async);
+}
+```
+
+We use the following helper APIs to indicate suspension points to the runtime, in addition to the runtime async call syntax:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// These methods are used to await things that cannot use runtime async signature form
+// TODO: Clarify which of these should be preferred? Should we always emit the `Unsafe` version when awaiting something that implements `ICriticalNotifyCompletion`?
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeHelpers
+{
+    [RuntimeAsyncMethod]
+    public static Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion;
+    [RuntimeAsyncMethod]
+    public static Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion;
+}
+```
+
+Additionally, we use the following helper attributes to indicate information to the runtime. If these attributes are not present in the reference assemblies, we will generate them; the runtime matches by full
+name, not by type identity, so we do not need to care about using the "canonical" versions.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// Used to tell the runtime to generate the async state machinery for this method
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodAttribute() : Attribute();
+
+// Used to mark locals that should be hoisted to the generated async closure. Note that the runtime does not guarantee that all locals marked with this attribute will be hoisted; if it can prove that it
+// doesn't need to hoist a variable, it may avoid doing so.
+public class HoistedLocal();
+```
+
+For experimentation purposes, we recognize an attribute that can be used to force the compiler to generate the runtime async code, or to force the compiler to generate a full state machine. This attribute is not
+defined in the BCL, and exists as an escape hatch for experimentation. It may be removed when the feature ships in stable.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodGenerationAttribute(bool runtimeAsync) : Attribute();
+```
+
+## Transformation strategy
+
+As mentioned previously, we try to expose as little of this to initial binding as possible. The one major exception to this is our handling of the `RuntimeAsyncMethodAttribute`; we do not let this be applied to
+user code, and will issue an error if a user tries to do this by hand.
+
+Compiler generated async state machines and runtime generated async share some of the same building blocks. Both need to have `await`s with in `catch` and `finally` blocks rewritten to pend the exceptions,
+perform the `await` outside of the `catch`/`finally` region, and then have the exceptions restored as necessary.
```

Comment:
@jakobbotsch this is the equivalent straight-line code. I think you're saying that those `ldflda` calls will need to be redone across the suspension point?

```il
        IL_0000: ldarg.0
        IL_0001: ldflda valuetype S C::S
        IL_0006: ldflda int32 S::X
        IL_000b: dup
        IL_000c: ldind.i4
        IL_000d: call int32 Program::Bar()
        IL_0012: add
        IL_0013: stind.i4
        IL_0014: ret
```

Reply:
From what I recall, a good way to see the ref scenario is `localArray[B] += await C;`. In normal code, we save the ref from indexing into array, but in async code, we'll obtain that ref before the suspension for `await` and again after resuming.

[sharplab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5lQDMmM6AwugN7Lq2bFQAcmAbOgLIAUAlOjXdSR1h6AJYA7AC4BtALroAhugC86cQFMA7mKlyq6AL74hI2gHoz6ACoALdTuDqAHhIDmOyQHtFAJx8KATzEAZ3R1ADcFABsAVwVJdWB0SU1RAGN7flMLdGCJDPRNew1E5Lt0H3UAM3QAI2rPSsVxJIUqhJ8y+2CY4IAHdXE8z3E6mMl0NIVxcU8Jm09RYInRSSyRBWl2GB55MFUoAE5WAB4JSQA+ADoAMR9PAFsAJXUeqMlOABYYbmNhA3WREwcDY5w4O14lEwAHZ0HBcIZkAYgA=)
```csharp
using System;
using System.Threading.Tasks;
public class C {
    public async Task M() 
    {
        int[] a = new int[] { };
        // The indexing into array is evaluated twice 
        // since we need the ref before and after the suspension but cannot hoist it
        a[M2()] += await Task<int>.FromResult(42);
    }
    public static int M2() { return 1; }
}
```

Reply:
> @jakobbotsch this is the equivalent straight-line code. I think you're saying that those `ldflda` calls will need to be redone across the suspension point?
> 
> ```
>         IL_0000: ldarg.0
>         IL_0001: ldflda valuetype S C::S
>         IL_0006: ldflda int32 S::X
>         IL_000b: dup
>         IL_000c: ldind.i4
>         IL_000d: call int32 Program::Bar()
>         IL_0012: add
>         IL_0013: stind.i4
>         IL_0014: ret
> ```

Right, exactly. There are more complex cases like @jcouv pointed out, or even impossible cases like `SomethingThatReturnsRefInt() += await Foo()` (this raises a compiler error today).